virtool/spaces/utils.py

```diff
@@ -0,0 +1,117 @@
+from datetime import datetime
+from typing import List, Union, Optional
+
+from virtool_core.models.basemodel import BaseModel
+from virtool_core.models.roles import (
+    SpaceRoleType,
+    SpaceRole,
+    SpaceLabelRole,
+    SpaceProjectRole,
+    SpaceReferenceRole,
+    SpaceSampleRole,
+    SpaceSubtractionRole,
+    SpaceUploadRole,
+)
+from virtool_core.models.user import UserNested
+from virtool_core.utils import document_enum
+
+from virtool.authorization.client import AuthorizationClient
+from virtool.authorization.relationships import UserRoleAssignment
+
+
+class SpaceMember(UserNested):
+    role: SpaceRole
+    label: Optional[SpaceLabelRole]
+    project: Optional[SpaceProjectRole]
+    reference: Optional[SpaceReferenceRole]
+    sample: Optional[SpaceSampleRole]
+    subtraction: Optional[SpaceSubtractionRole]
+    upload: Optional[SpaceUploadRole]
+
+
+class SpaceNested(BaseModel):
+    id: int
+    name: str
+
+
+class SpaceMinimal(SpaceNested):
+    description: str
+
+
+class Space(SpaceMinimal):
+    created_at: datetime
+    updated_at: datetime
+    created_by: str
+    members: List[SpaceMember]
+
+
+class SpaceSearchResult(BaseModel):
+    items: List[SpaceMinimal]
+    available_roles: List[dict]
+
+
+ENUM_LIST = [
+    SpaceRole,
+    SpaceLabelRole,
+    SpaceProjectRole,
+    SpaceReferenceRole,
+    SpaceSampleRole,
+    SpaceSubtractionRole,
+    SpaceUploadRole,
+]
+
+AVAILABLE_ROLES = [
+    {"id": role, "name": role.capitalize(), "description": role.__doc__}
+    for enum in ENUM_LIST
+    for role in document_enum(enum)
+]
+
+
+async def remove_user_roles(
+    authorization_client: AuthorizationClient,
+    user_id: Union[str, int],
+    space_id: int,
+    enum_list: List[SpaceRoleType],
+):
+
+    for enum_type in enum_list:
+        for role in enum_type:
+            await authorization_client.remove(
+                UserRoleAssignment(user_id, space_id, enum_type(role))
+            )
```

Comment:
I think these needs a more elegant solution.

The `ENUM_LIST` constant naming is not good. Two types mushed together...

Can you not transiently define `ENUM_LIST` as a tuple when you are creating `AVAILABLE_ROLES`. Then, it looks like you can iterate through available roles in `remove_user_roles` instead of `ENUM_LIST`.

Alternatively, `ENUM_LIST` needs a rename.

Let me know if you want to discuss.